File: src/main/java/frc/team4069/robot/commands/OperatorDriveCommand.java
package frc.team4069.robot.commands;

import frc.team4069.robot.io.Input;
import frc.team4069.robot.subsystems.DriveBaseSubsystem;

// The main command for operator control of the drive base
class OperatorDriveCommand extends CommandBase {

    // The distance that the current quick turn should travel (0 means quick turn is not being used)
    private double quickTurnDistanceMeters = 0;

    // Constructor, used to claim subsystems
    OperatorDriveCommand() {
        // Claim exclusive use of the drive base
        requires(driveBase);
    }

    // Called to initialize the drive base
    protected void initialize() {
        // It should start out idle
        driveBase.stop();
    }

    // Called frequently while this command is being run
    protected void execute() {
        // Set drive base speeds using the joystick inputs
        // The turning coefficient should be equal to the steering axis
        double turningCoefficient = Input.getSteeringAxis();
        // Use the negative of the joystick's speed axis as the speed of the drive base
        double speed = Input.getDriveSpeed();

        // If quick turn is not currently being used
        if (quickTurnDistanceMeters == 0) {
            // Get output from the directional pad
            int directionalPadAngle = Input.getDirectionalPadAngleDegrees();
            // If the directional pad is being used (the output is not -1) and the up direction is not
            // being pressed (an output of 0)
            if (directionalPadAngle != -1 && directionalPadAngle != 0) {
                // If the angle is greater than 180
                if (directionalPadAngle > 180) {
                    // Subtract 360 so that the robot rotates in the opposite direction
                    directionalPadAngle -= 360;
                }

                // Multiply the angle in rotations by the horizontal distance between the wheels
                // wheels in meters multiplied by pi, to get the length of the arc that the wheels should travel
                quickTurnDistanceMeters = (((double) directionalPadAngle) / 360)
                        * DriveBaseSubsystem.ROBOT_TRACK_WIDTH_METERS
                        * Math.PI;
                // Get the sign of the angle in order to calculate the direction to turn the wheels
                double turnDirection = Math.signum(quickTurnDistanceMeters);
                // Start turning at full speed in the direction of the sign
                driveBase.driveContinuousSpeed(turnDirection, 0);
            }
        }

        // If quick turn is currently running
        else {
            // If the distance travelled so far is greater than or equal to the quick turn distance
            if (driveBase.getDistanceTraveledMeters() >= Math.abs(quickTurnDistanceMeters)) {
                // Stop turning
                driveBase.stop();
                // Disable quick turn
                quickTurnDistanceMeters = 0;
            }
        }

        // If the driver is using the regular controls or if quick turn is disabled
        if (Math.abs(turningCoefficient) >= 0.2
                || Math.abs(speed) >= 0.2
                || quickTurnDistanceMeters == 0) {
            // Disable quick turn
            quickTurnDistanceMeters = 0;
            // Set the speed of the robot
            driveBase.driveContinuousSpeed(turningCoefficient, speed);
        }
    }

    // Called to check whether this command has completed
    protected boolean isFinished() {
        // Always return false because this command should never finish
        return false;
    }
}

File: src/main/java/frc/team4069/robot/commands/OperatorDriveCommand.kt
package frc.team4069.robot.commands

import edu.wpi.first.wpilibj.command.Command
import frc.team4069.robot.io.Input
import frc.team4069.robot.subsystems.DriveBaseSubsystem

class OperatorDriveCommand : Command() {

    private var quickTurnDistanceMetres = 0.0

    init {
        requires(DriveBaseSubsystem)
    }

    override fun initialize() {
        // Drive base should start out idle
        DriveBaseSubsystem.stop()
    }

    override fun execute() {
        // Get values from the controller.
        val turningCoefficient = Input.steeringAxis
        val speed = Input.driveSpeed

        // If we aren't currently quick turning
        if (quickTurnDistanceMetres == 0.0) {
            var directionalPadAngle = Input.directionalPadAngleDegrees
            // If we're using the directional pad, and not pressing up
            if (directionalPadAngle != -1 && directionalPadAngle != 0) {
                if (directionalPadAngle > 180) {
                    // Subtract 360 to have the robot rotate in the opposite direction
                    directionalPadAngle -= 360
                }

                // Get the arc that the wheels should travel
                quickTurnDistanceMetres = (directionalPadAngle.toDouble() / 360) * DriveBaseSubsystem.ROBOT_TRACK_WIDTH_METRES * Math.PI
                // Get the sign of the angle, then quick turn
                val turnDirection = Math.signum(quickTurnDistanceMetres)

                DriveBaseSubsystem.quickTurn(turnDirection)
            }
        } else { // Quick turn is running
            if (DriveBaseSubsystem.distanceTraveledMetres >= Math.abs(quickTurnDistanceMetres)) {
                DriveBaseSubsystem.stop()
                quickTurnDistanceMetres = 0.0
            }
        }

        // Driver is using regular controls
        if (Math.abs(turningCoefficient) >= 0.2
                || Math.abs(speed) >= 0.2
                || quickTurnDistanceMetres == 0.0) {
            DriveBaseSubsystem.drive(turningCoefficient, speed)
        }
    }

    // This command will never finish
    override fun isFinished() = false
}